[RISCV][TTI] Implement cost of intrinsic active_lane_mask

arcbbb opened the pull request:
This patch uses the argument type to infer the LMUL cost for the index generation, add, and comparison.

arcbbb committed:
[RISCV][TTI] Implement cost of intrinsic active_lane_mask

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -863,6 +863,25 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+          ICA.getArgTypes()[0], cast<VectorType>(RetTy)->getElementCount());
+      auto LT = getTypeLegalizationCost(ExpRetTy);
+
+      // vid.v   v8
```
`vid.v v8` will always be hoisted I think, should we take this into consideration?

arcbbb replied:
I thought for that case we may need an argument to know if the user want to include `vid.v` in the cost.
However upon reviewing BasicTTIImpl.h, I notice that the cost is calculated as a sum of  `uadd_sat` and `ICmp`.
I think I can follow that approach.

arcbbb committed:
Address comments

remove vid.v as BasicTTIImpl.h only calculates cost as sum of `uadd_sat` and
ICmp.

topperc commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -863,6 +863,21 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+          ICA.getArgTypes()[0], cast<VectorType>(RetTy)->getElementCount());
+      auto LT = getTypeLegalizationCost(ExpRetTy);
+
+      // vid.v   v8  // considered hoisted
```
Don't we need to count it for code size?

arcbbb replied:
This poses a dilemma as both statements hold true:
1) the `vid.v` would be hoisted.
2) BasicImpl doesn't count it for all CostKind.

so I am leaning toward to consistency with the base implementation.

topperc replied:
ok

topperc reviewed (approved):
LGTM
